src/QMCWaveFunctions/SpinorSet.cpp

```diff
@@ -211,6 +211,75 @@ void SpinorSet::mw_evaluateVGLWithSpin(const RefVectorWithLeader<SPOSet>& spo_li
   }
 }
 
+void SpinorSet::mw_evaluateVGLandDetRatioGradsWithSpin(const RefVectorWithLeader<SPOSet>& spo_list,
+                                                       const RefVectorWithLeader<ParticleSet>& P_list,
+                                                       int iat,
+                                                       const std::vector<const ValueType*>& invRow_ptr_list,
+                                                       OffloadMWVGLArray& phi_vgl_v,
+                                                       std::vector<ValueType>& ratios,
+                                                       std::vector<GradType>& grads,
+                                                       std::vector<ValueType>& spingrads) const
+{
+  auto& spo_leader = spo_list.getCastedLeader<SpinorSet>();
+  auto& P_leader   = P_list.getLeader();
+  assert(this == &spo_leader);
+  assert(phi_vgl_v.size(0) == DIM_VGL);
+  assert(phi_vgl_v.size(1) == spo_list.size());
+  const size_t nw             = spo_list.size();
+  const size_t norb_requested = phi_vgl_v.size(2);
+
+  OffloadMWVGLArray up_phi_vgl_v, dn_phi_vgl_v;
+  up_phi_vgl_v.resize(DIM_VGL, nw, norb_requested);
+  dn_phi_vgl_v.resize(DIM_VGL, nw, norb_requested);
+  std::vector<ValueType> up_ratios(nw), dn_ratios(nw);
+  std::vector<GradType> up_grads(nw), dn_grads(nw);
```

Comment:
Keep them as members. Resizing is quite expensive especially OffloadMWVGLArray.

Reply:
discussion on slack you mentioned to leave them as local for now. So I didn't touch this